src/libraries/System.Text.Json/System.Text.Json.SourceGeneration/JsonSourceGenerator.cs

```diff
@@ -16,47 +19,175 @@ namespace System.Text.Json.SourceGeneration
         }
 
-        // Temporary function for now that reads all types. Should search types with attribute JsonSerializable.
-        internal class JsonSerializableSyntaxReceiver : ISyntaxReceiver
+        public class JsonSerializableSyntaxReceiver : ISyntaxReceiver
         {
-            public List<TypeDeclarationSyntax> GeneratorInputTypes = new List<TypeDeclarationSyntax>();
+            public List<KeyValuePair<string, IdentifierNameSyntax>> ExternalClassTypes = new List<KeyValuePair<string, IdentifierNameSyntax>>();
+            public List<KeyValuePair<string, TypeDeclarationSyntax>> InternalClassTypes = new List<KeyValuePair<string, TypeDeclarationSyntax>>();
 
             public void OnVisitSyntaxNode(SyntaxNode syntaxNode)
             {
-                // Get all the type decl in all syntax tree.
-                if (syntaxNode is TypeDeclarationSyntax tds)
+                // Look for classes or structs for JsonSerializable Attribute.
+                if (syntaxNode is ClassDeclarationSyntax || syntaxNode is StructDeclarationSyntax)
                 {
-                    GeneratorInputTypes.Add(tds);
+                    // Find JsonSerializable Attributes.
+                    IEnumerable<AttributeSyntax>? serializableAttributes = null;
+                    AttributeListSyntax attributeList = ((TypeDeclarationSyntax)syntaxNode).AttributeLists.SingleOrDefault();
+                    if (attributeList != null)
+                    {
+                        serializableAttributes = attributeList.Attributes.Where(node => (node is AttributeSyntax attr && attr.Name.ToString() == "JsonSerializable")).Cast<AttributeSyntax>();
+                    }
+
+                    if (serializableAttributes?.Any() == true)
+                    {
+                        // JsonSerializableAttribute has AllowMultiple as False, should only have 1 attribute.
+                        Debug.Assert(serializableAttributes.Count() == 1);
+                        AttributeSyntax attributeNode = serializableAttributes.First();
+
+                        // Check if the attribute is being passed a type.
+                        if (attributeNode.DescendantNodes().Where(node => node is TypeOfExpressionSyntax).Any())
+                        {
+                            // Get JsonSerializable attribute arguments.
+                            AttributeArgumentSyntax attributeArgumentNode = (AttributeArgumentSyntax)attributeNode.DescendantNodes().Where(node => node is AttributeArgumentSyntax).SingleOrDefault();
+                            // Get external class token from arguments.
+                            IdentifierNameSyntax externalTypeNode = (IdentifierNameSyntax)attributeArgumentNode?.DescendantNodes().Where(node => node is IdentifierNameSyntax).SingleOrDefault();
+                            ExternalClassTypes.Add(new KeyValuePair<string, IdentifierNameSyntax>(((TypeDeclarationSyntax)syntaxNode).Identifier.Text, externalTypeNode));
```

Comment:
Please create an issue to ensure that this works when a custom attribute has behavior different from what we are expecting as described here: https://github.com/dotnet/runtimelab/pull/18#discussion_r458171580.

Reply:
https://github.com/dotnet/runtimelab/issues/29
closing issues